File: src/main/python/net/x3djsonld/Library/Tests/Components/ParticleSystems/Bubbles.py
####################################################################################################
#
# Now available: developmental python x3d.py package on PyPi for import.
#   This approach simplifies Python X3D deployment and use.
#   https://pypi.org/project/x3d
#
# Installation:
#       pip install x3d
# or
#       python -m pip install x3d
#
# Developer options for loading x3d package:
#
#    from x3d import *  # preferred approach, terser source that avoids x3d.* class prefixes
#
# or
#    import x3d         # traditional way to subclass x3d package, all classes require x3d.* prefix,
#                       # but python source is very verbose, for example x3d.Material x3d.Shape etc.
#                       # X3dToPython.xslt stylesheet insertPackagePrefix=true supports this option.
#
####################################################################################################

from x3d import *

newModel=X3D(profile='Full',version='3.3',
  head=head(
    children=[
    meta(name='comment',content='World of Titania'),
    meta(name='created',content='Sun, 15 May 2016 00:16:04 GMT'),
    meta(name='creator',content='holger'),
    meta(name='generator',content='Titania V2.0.0, http://titania.create3000.de'),
    meta(name='identifier',content='file:///home/holger/Projekte/Cobweb/excite/tests/Components/ParticleSystems/Bubbles.x3d'),
    meta(name='modified',content='Mon, 26 Sep 2016 02:51:17 GMT')]),
  Scene=Scene(
    children=[
    WorldInfo(title='Bubbles',
      metadata=MetadataSet(DEF='Titania',name='Titania',reference='http://titania.create3000.de',
        value=[
        MetadataSet(DEF='NavigationInfo',name='NavigationInfo',reference='http://titania.create3000.de',
          value=[
          MetadataString(DEF='type',name='type',reference='http://titania.create3000.de',value=["EXAMINE"])]),
        MetadataSet(DEF='Viewpoint',name='Viewpoint',reference='http://titania.create3000.de',
          value=[
          MetadataDouble(DEF='position',name='position',reference='http://titania.create3000.de',value=[1.62303498729208,40.0774585689728,42.6164648005402]),
          MetadataDouble(DEF='orientation',name='orientation',reference='http://titania.create3000.de',value=[0.948208773060804,-0.218726178841849,-0.230345352416253,5.85764824335848]),
          MetadataDouble(DEF='centerOfRotation',name='centerOfRotation',reference='http://titania.create3000.de',value=[0,0,0])]),
        MetadataSet(DEF='Selection',name='Selection',reference='http://titania.create3000.de',
          value=[
          MetadataSet(DEF='previous',name='previous',reference='http://titania.create3000.de'),
          MetadataSet(DEF='children',name='children',reference='http://titania.create3000.de',
            value=[
            DirectionalLight(DEF='_1',direction=(-0.372556,-3.8217e-05,-0.92801),shadows=True,shadowIntensity=0.518717)])])])),
    Background(DEF='Gray',skyColor=[(0.2,0.2,0.2)]),
    NavigationInfo(headlight=False),
    Viewpoint(description='Viewpoint',position=(0,20,55)),
    Transform(DEF='IndexedLineSet',
      children=[
      Shape(
        geometry=IndexedLineSet(colorPerVertex=False,colorIndex=[0,1,2],coordIndex=[0,1,-1,2,3,-1,4,5,-1],
          color=Color(color=[(1,0,0),(0,1,0),(0,0,1)]),
          coord=Coordinate(point=[(-1,0,0),(1,0,0),(0,-1,0),(0,1,0),(0,0,-1),(0,0,1)])))]),
    ParticleSystem(geometryType='GEOMETRY',maxParticles=500,particleLifetime=8,lifetimeVariation=0.8,particleSize=(1,1),
      emitter=PointEmitter(direction=(0,0,0),speed=0.5,variation=0.5,mass=0.1,surfaceArea=0.1),
      physics=WindPhysicsModel(direction=(0,1,0),speed=0.6,gustiness=1.7,turbulence=0.3),
      WindPhysicsModel(speed=1,gustiness=1.5)
      appearance=Appearance(
        material=Material(ambientIntensity=0.25,diffuseColor=(1,0.520125,0.73699),specularColor=(0.951515,0.951515,0.951515),shininess=0.557576)),
      geometry=Sphere(),),
    Transform(DEF='Rectangle2D',translation=(0,19.2591,-14.5791),scale=(43.4935,20.2591,1),
      children=[
      Shape(
        appearance=Appearance(
          material=Material(),),
        geometry=Rectangle2D(),)]),
    DirectionalLight(USE='_1'),
    Script(DEF='PhongShadingScript')])
) # X3D model complete

####################################################################################################
# Self-test diagnostics
####################################################################################################

print('Self-test diagnostics:')
if        metaDiagnostics(newModel): # built-in utility method in X3D class
    print(metaDiagnostics(newModel)) # display meta info, hint, warning, error, TODO values in this model
# print('check newModel.XML() serialization...')
newModelXML= newModel.XML() # test export method XML() for exceptions during export
newModel.XMLvalidate()
# print(newModelXML) # diagnostic

try:
#   print('check newModel.VRML() serialization...')
    newModelVRML=newModel.VRML() # test export method VRML() for exceptions during export
    # print(prependLineNumbers(newModelVRML)) # debug
    print("Python-to-VRML export of VRML output successful", flush=True)
except Exception as err: # usually BaseException
    # https://stackoverflow.com/questions/18176602/how-to-get-the-name-of-an-exception-that-was-caught-in-python
    print("*** Python-to-VRML export of VRML output failed:", type(err).__name__, err)
    if newModelVRML: # may have failed to generate
        print(prependLineNumbers(newModelVRML, err.lineno))

try:
#   print('check newModel.JSON() serialization...')
    newModelJSON=newModel.JSON() # test export method JSON() for exceptions during export
#   print(prependLineNumbers(newModelJSON)) # debug
    print("Python-to-JSON export of JSON output successful (under development)")
except Exception as err: # usually SyntaxError
    print("*** Python-to-JSON export of JSON output failed:", type(err).__name__, err)
    if newModelJSON: # may have failed to generate
        print(prependLineNumbers(newModelJSON,err.lineno))

print("python load and self-test diagnostics complete.")

File: src/main/c/net/coderextreme/Library/Tests/Components/ParticleSystems/Bubbles.c
#include "C:/x3d-code/www.web3d.org/x3d/languages/c/Concretes.h"
void main(int argc, char ** argv) {
Browser browser = X3D.getBrowser();
X3D X3D0;
X3D0.profile = "Full";
X3D0.version = "3.3";
head head1 = createNode("head");
meta meta2 = createNode("meta");
meta2.name = "comment";
meta2.content = "World of Titania";
head1.meta = new MFNode();

head1.meta[0] = meta2;

meta meta3 = createNode("meta");
meta3.name = "created";
meta3.content = "Sun, 15 May 2016 00:16:04 GMT";
head1.meta[1] = meta3;

meta meta4 = createNode("meta");
meta4.name = "creator";
meta4.content = "holger";
head1.meta[2] = meta4;

meta meta5 = createNode("meta");
meta5.name = "generator";
meta5.content = "Titania V2.0.0, http://titania.create3000.de";
head1.meta[3] = meta5;

meta meta6 = createNode("meta");
meta6.name = "identifier";
meta6.content = "file:///home/holger/Projekte/Cobweb/excite/tests/Components/ParticleSystems/Bubbles.x3d";
head1.meta[4] = meta6;

meta meta7 = createNode("meta");
meta7.name = "modified";
meta7.content = "Mon, 26 Sep 2016 02:51:17 GMT";
head1.meta[5] = meta7;

head = head1;

WorldInfo WorldInfo9 = createNode("WorldInfo");
WorldInfo9.title = "Bubbles";
MetadataSet MetadataSet10 = createNode("MetadataSet");
MetadataSet10.name = "Titania";
MetadataSet10.DEF = "Titania";
MetadataSet10.reference = "http://titania.create3000.de";
MetadataSet MetadataSet11 = createNode("MetadataSet");
MetadataSet11.name = "NavigationInfo";
MetadataSet11.DEF = "NavigationInfo";
MetadataSet11.reference = "http://titania.create3000.de";
MetadataString MetadataString12 = createNode("MetadataString");
MetadataString12.name = "type";
MetadataString12.DEF = "type";
MetadataString12.reference = "http://titania.create3000.de";
MetadataString12.value = new MFString(new java.lang.String["EXAMINE"]);
MetadataSet11.value = new MFNode();

MetadataSet11.value[0] = MetadataString12;

MetadataSet10.value = new MFNode();

MetadataSet10.value[0] = MetadataSet11;

MetadataSet MetadataSet13 = createNode("MetadataSet");
MetadataSet13.name = "Viewpoint";
MetadataSet13.DEF = "Viewpoint";
MetadataSet13.reference = "http://titania.create3000.de";
MetadataDouble MetadataDouble14 = createNode("MetadataDouble");
MetadataDouble14.name = "position";
MetadataDouble14.DEF = "position";
MetadataDouble14.reference = "http://titania.create3000.de";
MetadataDouble14.value = new MFDouble(new double[1.62303498729208,40.0774585689728,42.6164648005402]);
MetadataSet13.value = new MFNode();

MetadataSet13.value[0] = MetadataDouble14;

MetadataDouble MetadataDouble15 = createNode("MetadataDouble");
MetadataDouble15.name = "orientation";
MetadataDouble15.DEF = "orientation";
MetadataDouble15.reference = "http://titania.create3000.de";
MetadataDouble15.value = new MFDouble(new double[0.948208773060804,-0.218726178841849,-0.230345352416253,5.85764824335848]);
MetadataSet13.value[1] = MetadataDouble15;

MetadataDouble MetadataDouble16 = createNode("MetadataDouble");
MetadataDouble16.name = "centerOfRotation";
MetadataDouble16.DEF = "centerOfRotation";
MetadataDouble16.reference = "http://titania.create3000.de";
MetadataDouble16.value = new MFDouble(new double[0,0,0]);
MetadataSet13.value[2] = MetadataDouble16;

MetadataSet10.value[1] = MetadataSet13;

MetadataSet MetadataSet17 = createNode("MetadataSet");
MetadataSet17.name = "Selection";
MetadataSet17.DEF = "Selection";
MetadataSet17.reference = "http://titania.create3000.de";
MetadataSet MetadataSet18 = createNode("MetadataSet");
MetadataSet18.name = "previous";
MetadataSet18.DEF = "previous";
MetadataSet18.reference = "http://titania.create3000.de";
MetadataSet17.value = new MFNode();

MetadataSet17.value[0] = MetadataSet18;

MetadataSet MetadataSet19 = createNode("MetadataSet");
MetadataSet19.name = "children";
MetadataSet19.DEF = "children";
MetadataSet19.reference = "http://titania.create3000.de";
DirectionalLight DirectionalLight20 = createNode("DirectionalLight");
DirectionalLight20.DEF = "_1";
DirectionalLight20.direction = new SFVec3f(new float[-0.372556,-0.000038217,-0.92801]);
DirectionalLight20.shadows = True;
DirectionalLight20.shadowIntensity = 0.518717;
MetadataSet19.value = DirectionalLight20;

MetadataSet17.value[1] = MetadataSet19;

MetadataSet10.value[2] = MetadataSet17;

WorldInfo9.metadata = MetadataSet10;

children = new MFNode();

children[0] = WorldInfo9;

Background Background21 = createNode("Background");
Background21.DEF = "Gray";
Background21.skyColor = new MFColor(new float[0.2,0.2,0.2]);
children[1] = Background21;

NavigationInfo NavigationInfo22 = createNode("NavigationInfo");
NavigationInfo22.headlight = False;
children[2] = NavigationInfo22;

Viewpoint Viewpoint23 = createNode("Viewpoint");
Viewpoint23.description = "Viewpoint";
Viewpoint23.position = new SFVec3f(new float[0,20,55]);
children[3] = Viewpoint23;

Transform Transform24 = createNode("Transform");
Transform24.DEF = "IndexedLineSet";
Shape Shape25 = createNode("Shape");
IndexedLineSet IndexedLineSet26 = createNode("IndexedLineSet");
IndexedLineSet26.colorPerVertex = False;
IndexedLineSet26.colorIndex = new MFInt32(new int[0,1,2]);
IndexedLineSet26.coordIndex = new MFInt32(new int[0,1,-1,2,3,-1,4,5,-1]);
Color Color27 = createNode("Color");
Color27.color = new MFColor(new float[1,0,0,0,1,0,0,0,1]);
IndexedLineSet26.color = Color27;

Coordinate Coordinate28 = createNode("Coordinate");
Coordinate28.point = new MFVec3f(new float[-1,0,0,1,0,0,0,-1,0,0,1,0,0,0,-1,0,0,1]);
IndexedLineSet26.coord = Coordinate28;

Shape25.geometry = IndexedLineSet26;

Transform24.child = new undefined();

Transform24.child[0] = Shape25;

children[4] = Transform24;

ParticleSystem ParticleSystem29 = createNode("ParticleSystem");
ParticleSystem29.geometryType = "GEOMETRY";
ParticleSystem29.maxParticles = 500;
ParticleSystem29.particleLifetime = 8;
ParticleSystem29.lifetimeVariation = 0.8;
ParticleSystem29.particleSize = new SFVec2f(new float[1,1]);
PointEmitter PointEmitter30 = createNode("PointEmitter");
PointEmitter30.direction = new SFVec3f(new float[0,0,0]);
PointEmitter30.speed = 0.5;
PointEmitter30.variation = 0.5;
PointEmitter30.mass = 0.1;
PointEmitter30.surfaceArea = 0.1;
ParticleSystem29.emitter = PointEmitter30;

WindPhysicsModel WindPhysicsModel31 = createNode("WindPhysicsModel");
WindPhysicsModel31.direction = new SFVec3f(new float[0,1,0]);
WindPhysicsModel31.speed = 0.6;
WindPhysicsModel31.gustiness = 1.7;
WindPhysicsModel31.turbulence = 0.3;
ParticleSystem29.physics = new MFNode();

ParticleSystem29.physics[0] = WindPhysicsModel31;

WindPhysicsModel WindPhysicsModel32 = createNode("WindPhysicsModel");
WindPhysicsModel32.speed = 1;
WindPhysicsModel32.gustiness = 1.5;
ParticleSystem29.physics[1] = WindPhysicsModel32;

Appearance Appearance33 = createNode("Appearance");
Material Material34 = createNode("Material");
Material34.ambientIntensity = 0.25;
Material34.diffuseColor = new SFColor(new float[1,0.520125,0.73699]);
Material34.specularColor = new SFColor(new float[0.951515,0.951515,0.951515]);
Material34.shininess = 0.557576;
Appearance33.material = Material34;

ParticleSystem29.appearance = Appearance33;

Sphere Sphere35 = createNode("Sphere");
ParticleSystem29.geometry = Sphere35;

children[5] = ParticleSystem29;

Transform Transform36 = createNode("Transform");
Transform36.DEF = "Rectangle2D";
Transform36.translation = new SFVec3f(new float[0,19.2591,-14.5791]);
Transform36.scale = new SFVec3f(new float[43.4935,20.2591,1]);
Shape Shape37 = createNode("Shape");
Appearance Appearance38 = createNode("Appearance");
Material Material39 = createNode("Material");
Appearance38.material = Material39;

Shape37.appearance = Appearance38;

Rectangle2D Rectangle2D40 = createNode("Rectangle2D");
Shape37.geometry = Rectangle2D40;

Transform36.child = new undefined();

Transform36.child[0] = Shape37;

children[6] = Transform36;

DirectionalLight DirectionalLight41 = createNode("DirectionalLight");
DirectionalLight41.USE = "_1";
children[7] = DirectionalLight41;

Script Script42 = createNode("Script");
Script42.DEF = "PhongShadingScript";

Script42.setSourceCode(`ecmascript:\n"+
"\n"+
"function initialize ()\n"+
"{\n"+
"	Browser .setBrowserOption (\"Shading\", \"PHONG\");\n"+
"}`)
children[8] = Script42;

}
